rust/noosphere-ns/src/name_system.rs

```diff
@@ -0,0 +1,195 @@
+    /// In seconds, the Time-To-Live (TTL) duration of records set
+    /// in the network, and implicitly, how often the records are
+    /// propagated.
+    pub(crate) ttl: u64,
+    /// Map of sphere DIDs to [NSRecord] hosted/propagated by this name system.
+    pub(crate) hosted_records: HashMap<String, NSRecord>,
+    /// Map of resolved sphere DIDs to resolved [NSRecord].
+    pub(crate) resolved_records: HashMap<String, NSRecord>,
+}
+
+impl<'a> NameSystem<'a> {
+    /// Initializes and attempts to connect to the network.
+    pub async fn connect(&mut self) -> Result<()> {
+        let mut dht = DHTNode::new(self.key_material, self.bootstrap_peers, &self.dht_config)?;
+        dht.run().map_err(|e| anyhow!(e.to_string()))?;
+        dht.bootstrap().await.map_err(|e| anyhow!(e.to_string()))?;
+        dht.wait_for_peers(1)
+            .await
+            .map_err(|e| anyhow!(e.to_string()))?;
+        self.dht = Some(dht);
+        Ok(())
+    }
+
+    /// Disconnect and deallocate connections to the network.
+    pub fn disconnect(&mut self) -> Result<()> {
+        if let Some(mut dht) = self.dht.take() {
+            dht.terminate()?;
+        }
+        Ok(())
+    }
+
+    /// Propagates all hosted records on nearby peers in the DHT network.
+    /// in the DHT network. Automatically called every `ttl` seconds (TBD),
+    /// but can be manually called push updated records to the network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn propagate_records(&self) -> Result<()> {
+        if self.dht.is_none() {
+            return Err(anyhow!("not connected"));
+        }
+        if self.hosted_records.is_empty() {
+            return Ok(());
+        }
+
+        let pending_tasks: Vec<_> = self
+            .hosted_records
+            .iter()
+            .map(|(identity, record)| self.dht_set_record(identity, record))
+            .collect();
+        try_join_all(pending_tasks).await?;
+        Ok(())
+    }
+
+    /// Propagates the corresponding managed sphere's content Cid on nearby peers
+    /// in the DHT network.
+    ///
+    /// Can fail if NameSystem is not connected or if no peers can be found.
+    pub async fn set_record(&mut self, identity: &String, link: &Cid) -> Result<()> {
+        if self.dht.is_none() {
+            return Err(anyhow!("not connected"));
+        }
+
+        let record = NSRecord::new_from_ttl(link.to_owned(), self.ttl)?;
+        self.dht_set_record(identity, &record).await?;
+        self.hosted_records.insert(identity.to_owned(), record);
+        Ok(())
+    }
+
+    /// Gets the content Cid for the provided sphere identity.
+    pub async fn get_record(&mut self, identity: &String) -> Option<&Cid> {
+        // Round about way of checking for local valid record before
+        // querying the DHT network due to the borrow checker.
+        // https://stackoverflow.com/questions/70779967/rust-borrow-checker-and-early-returns#
```

Comment:
Not ideal, but an early return always holds on to ownership, even when the conditional is not triggered. This work-around works, but ideally we don't have to get the record twice from the HashMap. Any ideas @cdata ?

Reply:
a Mutex seems like a heavy workaround for this, but there's probably some verbose simple solution as it seems to be a borrow checker "error"? (the hubris)

Reply:
The `Mutex` may be a heavy workaround, but also: `get_record` being a mutating operation is counter-intuitive (although perfectly understandable when reading the implementation).

No action required, just food for thought.